src/libpriv/rpmostree-kernel.c

```diff
@@ -211,52 +206,112 @@ rpmostree_find_kernel (int rootfs_dfd,
 {
-  g_autoptr(GChecksum) boot_checksum = NULL;
-  g_autofree char *kernel_final_path = NULL;
-  g_autofree char *initramfs_final_path = NULL;
-  const char *boot_checksum_str = NULL;
-
-  /* Now, calculate the combined sha256sum of the two. We checksum the initramfs
-   * from the tmpfile fd (via mmap()) to avoid writing it to disk in another
-   * temporary location.
+  const char slash_bootdir[] = "boot";
+  const char usrlibostree_bootdir[] = "usr/lib/ostree-boot";
+  g_autofree char *modules_bootdir = g_strconcat ("usr/lib/modules/", kver, NULL);
+
+  /* Calculate the sha256sum of the kernel+initramfs (called the "boot
+   * checksum"). We checksum the initramfs from the tmpfile fd (via mmap()) to
+   * avoid writing it to disk in another temporary location.
    */
-  boot_checksum = g_checksum_new (G_CHECKSUM_SHA256);
+  g_autoptr(GChecksum) boot_checksum = g_checksum_new (G_CHECKSUM_SHA256);
   if (!_rpmostree_util_update_checksum_from_file (boot_checksum, rootfs_dfd, kernel_path,
                                                   cancellable, error))
     return FALSE;
-
   { g_autoptr(GMappedFile) mfile = g_mapped_file_new_from_fd (initramfs_tmpf->fd, FALSE, error);
     if (!mfile)
       return FALSE;
     g_checksum_update (boot_checksum, (guint8*)g_mapped_file_get_contents (mfile),
                        g_mapped_file_get_length (mfile));
   }
-  boot_checksum_str = g_checksum_get_string (boot_checksum);
-
-  kernel_final_path = g_strconcat (bootdir, "/", "vmlinuz-", kver, "-", boot_checksum_str, NULL);
-  initramfs_final_path = g_strconcat (bootdir, "/", "initramfs-", kver, ".img-", boot_checksum_str, NULL);
+  const char *boot_checksum_str = g_checksum_get_string (boot_checksum);
 
-  /* Put the kernel in the final location */
-  if (!glnx_renameat (rootfs_dfd, kernel_path, rootfs_dfd, kernel_final_path, error))
-    return FALSE;
-  /* Link the initramfs directly to its final destination */
+  g_autofree char *kernel_modules_path = g_strconcat (modules_bootdir, "/vmlinuz", NULL);;
+  /* It's possible the bootdir is already the modules directory; in that case,
+   * we don't need to rename.
+   */
+  if (!g_str_equal (kernel_path, kernel_modules_path))
+    {
+      g_assert_cmpstr (bootdir, !=, modules_bootdir);
+      if (!glnx_renameat (rootfs_dfd, kernel_path, rootfs_dfd, kernel_modules_path, error))
+        return FALSE;
+    }
+  g_autofree char *initramfs_modules_path = g_strconcat (modules_bootdir, "/initramfs.img", NULL);
   if (!glnx_link_tmpfile_at (initramfs_tmpf, GLNX_LINK_TMPFILE_NOREPLACE,
-                             rootfs_dfd, initramfs_final_path,
+                             rootfs_dfd, initramfs_modules_path,
                              error))
     return FALSE;
 
+  /* Update /usr/lib/ostree-boot and /boot (if we find a kernel there) */
+  const char *legacy_paths[] = { usrlibostree_bootdir, slash_bootdir };
+  int upto_index;
+  switch (dest)
+    {
+    case RPMOSTREE_FINALIZE_KERNEL_AUTO:
```

Comment:
This confuses me; don't we always want to put the kernel in `/usr/lib/ostree-boot` as well?

Reply:
Yes; we do/should - this was more about "update /boot only if we find a kernel there".

Reply:
Ahh gotcha. OK, the implementation here is really tricky. Can we just drop the loop approach, factor out the `unlink` & `linkat` bits and just go through the two locations one at a time? Also, maybe a comment where the enum is declared for `AUTO`?

Reply:
OK, reworked, and fixed some bugs along the way.  Also did a `git add` of a test case I'd written that had been sitting in my tree for a while without me doing a `git clean -dfx` (lucky!).